test/type/self.swift

```diff
@@ -41,6 +41,13 @@ final class FinalMario : Mario {
     override func getFriend() -> FinalMario {
         return FinalMario()
     }
+
+  final func savePrincess(princess: Self) -> Self {
+    let staticPrincess: FinalMario = princess
+    let alsoStaticPrincess: Self = staticPrincess
+
+    return princess // expected-error {{cannot convert return expression of type 'FinalMario' to return type 'Self'}}
```

Comment:
This could be made to work by allowing the conversion constraint to be solved if the underlying type is final.